File: libwebrtc-sys/src/sdp_video_format.cc
#include "libwebrtc-sys/include/sdp_video_format.h"
#include "libwebrtc-sys/src/shared_bridge.rs.h"

rust::Vec<ArcasRustDict> sdp_video_format_get_parameters(const webrtc::SdpVideoFormat &format)
{
    rust::Vec<ArcasRustDict> list;

    for (auto param : format.parameters)
    {
        ArcasRustDict dict;
        dict.key = rust::String(param.first.c_str());
        dict.value = rust::String(param.second.c_str());
        list.push_back(dict);
    }
    return list;
}

const std::string &sdp_video_format_get_name(const webrtc::SdpVideoFormat &format)
{
    return format.name;
}

rust::String sdp_video_format_to_string(const webrtc::SdpVideoFormat &format)
{
    return rust::String(format.ToString().c_str());
}

std::unique_ptr<webrtc::SdpVideoFormat> create_sdp_video_format(ArcasSdpVideoFormatInit init)
{

    std::map<std::string, std::string> cxx_parameters;

    for (auto item : init.parameters)
    {
        cxx_parameters.insert(std::make_pair(std::string(item.key.c_str()), std::string(item.value.c_str())));
    }

    return std::make_unique<webrtc::SdpVideoFormat>(std::string(init.name.c_str()), cxx_parameters);
}

std::unique_ptr<std::vector<webrtc::SdpVideoFormat>> create_sdp_video_format_list(ArcasSdpVideoFormatVecInit list)
{
    auto output = std::make_unique<std::vector<webrtc::SdpVideoFormat>>();

    for (auto rust_sdp_video_format : list.list)
    {
        std::map<std::string, std::string> cxx_parameters;

        for (auto item : rust_sdp_video_format.parameters)
        {
            cxx_parameters.insert(std::make_pair(std::string(item.key.c_str()), std::string(item.value.c_str())));
        }
        output->push_back(webrtc::SdpVideoFormat(std::string(rust_sdp_video_format.name.c_str()), cxx_parameters));
    }

    return output;
}

File: libwebrtc-sys/src/sdp_video_format.rs
#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("include/alias.h");
        include!("include/sdp_video_format.h");
        include!("libwebrtc-sys/src/shared_bridge.rs.h");
        type ArcasCxxSdpVideoFormat = crate::shared_bridge::ffi::ArcasCxxSdpVideoFormat;
        type ArcasRustDict = crate::shared_bridge::ffi::ArcasRustDict;
        type ArcasSdpVideoFormatInit = crate::shared_bridge::ffi::ArcasSdpVideoFormatInit;
        type ArcasSdpVideoFormatVecInit = crate::shared_bridge::ffi::ArcasSdpVideoFormatVecInit;

        // ArcasCxxSdpVideoFormat
        #[cxx_name = "IsSameCodec"]
        unsafe fn is_same_codec(
            self: &ArcasCxxSdpVideoFormat,
            other: &ArcasCxxSdpVideoFormat,
        ) -> bool;

        #[cxx_name = "sdp_video_format_get_parameters"]
        fn video_format_get_parameters(format: &ArcasCxxSdpVideoFormat) -> Vec<ArcasRustDict>;

        #[cxx_name = "sdp_video_format_get_name"]
        fn sdp_video_format_get_name(format: &ArcasCxxSdpVideoFormat) -> &'static CxxString;

        #[cxx_name = "sdp_video_format_to_string"]
        fn sdp_video_format_to_string(format: &ArcasCxxSdpVideoFormat) -> String;

        // Helper for use with the VideoFactory methods but generically useful too
        //
        // Where optional types are used this is also how we pass "none" (empty vec)
        // These types are relevant:
        //
        //  - absl::optional<SdpVideoFormat>
        //  - std::vector<SdpVideoFormat>
        //
        fn create_sdp_video_format_list(
            format_list: ArcasSdpVideoFormatVecInit,
        ) -> UniquePtr<CxxVector<ArcasCxxSdpVideoFormat>>;

        fn create_sdp_video_format(
            init: ArcasSdpVideoFormatInit,
        ) -> UniquePtr<ArcasCxxSdpVideoFormat>;
    }
}
